Allow identity templates in ssh backend `default_user` field

ianferguson opened the pull request:
Fixes #16350 by enabling Identity attribute templating support for the `default_user` field on the SSH backend roles

stevendpclark commented on builtin/logical/ssh/backend_test.go:
```diff
@@ -340,6 +340,147 @@ func TestBackend_AllowedUsersTemplate_WithStaticPrefix(t *testing.T) {
+		t.Errorf("signing request should fail when default_user is not in the allowed_users list, because allowed_users_template is true and default_user_template is not")
+	}
+
+	expectedErrStr := ":TKTKT"
```
Could we fix this test as it's failing now. I get the impression this expected error string was meant to be updated but was missed?

ianferguson replied:
it was, sorry about that! I've committed the expected value in 6ecf401852

stevendpclark commented on builtin/logical/ssh/backend_test.go:
```diff
@@ -1893,6 +2034,66 @@ func getSshCaTestCluster(t *testing.T, userIdentity string) (*vault.TestCluster,
 	return cluster, userpassToken
 }
 
+// TKTK
```
Could this comment be removed please?

ianferguson replied:
Sorry about that! I've removed it in 6ecf401852

stevendpclark reviewed (changes requested):
Thanks @ianferguson for the submission! 

Overall this looks great! I do have a few little comments to bring this across the line, see individual other comments. 

Additionally  an update to the docs for the new field/behavior within [ssh.mdx](https://github.com/hashicorp/vault/blob/main/website/content/api-docs/secret/ssh.mdx) would be needed if you don't mind?

ianferguson commented:
@stevendpclark Thanks for the review - I've addressed the comments (sorry about both of those!) in 6ecf401852 and added the requested docs in 44ad5d49e2

cipherboy commented on builtin/logical/ssh/path_issue_sign.go:
```diff
@@ -150,23 +174,12 @@ func (b *backend) calculateValidPrincipals(data *framework.FieldData, req *logic
-				allowedPrincipals = append(allowedPrincipals, principal)
+			rendered, err := b.renderPrincipal(principal, req)
+			if err != nil {
+				return nil, fmt.Errorf("template '%s' could not be rendered -> %s", principal, err)
```
This is inconsistent with above (lines 77) -- I believe the err should be returned here and this message should be preserved as-is as we already contain a copy of the error message prefix you have here (so it'll be duplicated). 

IOW, I think 77 is wrong too (`return nil, err`) to preserve existing behavior-ish.

stevendpclark replied:
Oops good catch @cipherboy. I'll merge this PR and address this in another myself.

ianferguson replied:
previously (when only the allowed_users field could be templated) the template rendering happened in `calculateValidPrincipals` and the error on this line (`return nil, fmt.Errorf("template '%s' could not be rendered -> %s", principal, err)`, on line 163 in the previous version) could be returned by that func, 

That return value would then be caught on the previous version's line 74 which contained the logical.ErrorResponse wrapping, i.e. lines 73-76 prior to my changes were:
```
parsedPrincipals, err = b.calculateValidPrincipals(data, req, role, role.DefaultUser, role.AllowedUsers, strutil.StrListContains)
if err != nil {
	return logical.ErrorResponse(err.Error()), nil
} 
```

I was aiming to match the failure behaviors for templating errors in `default_user` to the existing behavior for `allowed_users` -- that all said, :+1: with any changes to it, y'all know the details on that better than me!

stevendpclark commented:
Thanks @ianferguson for the contribution!
